pkg/cloud/api/mutable_resource.go

```diff
@@ -272,24 +291,51 @@ func (u *mutableResource[GA, Alpha, Beta]) AccessBeta(f func(x *Beta)) error {
 	return u.postAccess(meta.VersionBeta, 0)
 }
 
+// completeType means that the object does not have conversion errors and its
+// type is not missing.
+func completeType(err error, obj interface{}) bool {
+	return err == nil && !IsTypeMissing(obj)
+}
+
+// convertibleTo means that the object does not have conversion errors or its
+// type is missing.
+func convertibleTo(err error, obj interface{}) bool {
+	return err == nil || IsTypeMissing(obj)
+}
+
+// convertibleTo means that the object has conversion errors or its type is
+// missing.
+func notConvertibleTo(err error, obj interface{}) bool {
+	return err != nil || IsTypeMissing(obj)
+}
+
 func (u *mutableResource[GA, Alpha, Beta]) ImpliedVersion() (meta.Version, error) {
 	_, gaErr := u.ToGA()
 	_, alphaErr := u.ToAlpha()
 	_, betaErr := u.ToBeta()
 
-	switch {
-	case gaErr == nil && alphaErr == nil && betaErr == nil:
+	// If type is GA and converts to Alpha and Beta without error return GA.
+	// If type Alpha or Beta has missing type ignore conversion errors.
+	if completeType(gaErr, u.ga) &&
+		convertibleTo(alphaErr, u.alpha) &&
+		convertibleTo(betaErr, u.beta) {
 		return meta.VersionGA, nil
-	case gaErr != nil && alphaErr == nil && betaErr != nil:
+	}
+
+	switch {
+	case completeType(alphaErr, u.alpha) && notConvertibleTo(betaErr, u.beta):
 		return meta.VersionAlpha, nil
-	case gaErr != nil && alphaErr != nil && betaErr == nil:
+	case notConvertibleTo(alphaErr, u.alpha) && completeType(betaErr, u.beta):
 		return meta.VersionBeta, nil
 	default:
 		return meta.VersionGA, fmt.Errorf("indeterminant version (ga=%v, alpha=%v, beta=%v)", gaErr, alphaErr, betaErr)
 	}
 }
 
 func (u *mutableResource[GA, Alpha, Beta]) ToGA() (*GA, error) {
+	if IsTypeMissing(u.ga) {
+		return nil, &MissingTypeError{}
```

Comment:
See comment above -- we should return a better error, otherwise people will end up with a not-so- useful error message in their logs.

Reply:
changed